### PR TITLE
WIP classify to rgba

### PR DESCRIPTION
**Pull request by knaaptime**
this adds a function to generate an array of colors from an array of values. I've [been toying](https://knaaptime.com/urban_analysis/05_access/simple_access.html#using-the-z-dimension) with pydeck lately, which, basically, requires a column of colors on the df. Looking a bit closer, that's how most (all?) of the viz libraries work, so this is a shot at refactoring geopandas's logic for generating colors.

I can add tests, but wanted to open first to see if folks find this valuable and if it should live here

### CODE REVIEW

**Comment by martinfleis** [comment]
I am think whether this should not be a class or a method on `MapClassifier` (probably the latter). Because to use it with a legend, you will normally be interested in other information apart from the array of colours.

**Comment by knaaptime** [comment]
i was curious about that, but it seemed like classify was doing the [heavy lifting](https://github.com/geopandas/geopandas/blob/5186e69d9b81b2bd73fce1d85f3852971d227f8e/geopandas/plotting.py#L714) in geopandas, even with the legend?

probably no real reason we couldn't do both. The difference is a classifier class doesn't require a cmap

**Comment by knaaptime** [comment]
ah. but `classify` gives back the fitted classifier whereas this gives back the array.

Agree this makes more sense as a method, will move around

**Comment by knaaptime** [comment]
nevermind. If this gets pushed inside the class, there's nowhere to include the nan logic

**Comment by martinfleis** [comment]
That is a bit unfortunate to do the classification twice. Especially given some of them can be quite costly. Maybe include NaN tracking in the class? Like your `legit_indices` here?

**Comment by knaaptime** [comment]
yeah, agree. I think the nan-handle logic is only like those three lines. Without touching any of the classification code, we could maybe sneak it into the [first step](https://github.com/pysal/mapclassify/blob/d9c60ce2dc5ea251b87824b5f53b32b5bcb45c86/mapclassify/classifiers.py#L320) of the binning function so nans are ignored from the outset?

File: mapclassify/_classify_API.py [hunk 1]
```diff
@@ -214,3 +218,60 @@ def classify(
...
+
+    # only operate on non-NaN values
+    v = pd.Series(values)
+    legit_indices = v[~v.isna()].index.values
+
+    # transform (non-NaN) values into class bins
+    bins = classify(v.dropna().values, scheme=classifier, k=k).yb
+
+    # create a normalizer using the data's range (not strictly 1-k...)
+    norm = Normalize(min(bins), max(bins))
```
Comment (sjsrey) [inline mapclassify/_classify_API.py:258]: I'm not clear why there is a classifier called prior to here. 
It seems to me this is trying to do a classless choropleth map [1].
If that is true, then the first classifier can be omitted and line 258 could become
`norm = Normalize(v.min(), v.max())`

But maybe I'm missing something here in my understanding?


[1] Tobler, W. (1973). Choropleth maps without class intervals. Geographical
Analysis, 5:262-265

Comment (sjsrey) [reply mapclassify/_classify_API.py:258]: image
So do I have this right? 

This was classified k=8, fisher jenks. So I'm thinking there should only be 8 colors in the figure, but the additional z axis allows for differentiation between units in the same bin but with different values. The color of the hex fill would be whatever class the hex value is placed into.

Comment (knaaptime) [reply mapclassify/_classify_API.py:258]: Z and color are disconnnected there. Color is based on Jenna but height is continuous

Comment (knaaptime) [reply mapclassify/_classify_API.py:258]: (i.e. all this does is take a cmap and translate bins --> colors)

so if you remove the z-dimension, this function gives you the same choropleth geopandas would give you. But then you can also use extrusion to encode the same or a different variable. A bit more at the [end of this](https://knaaptime.com/urban_analysis/02_geoprocessing/01_geovisualization.html#pydeck)

**Comment by sjsrey** [comment]
> this adds a function to generate an array of colors from an array of values. I've [been toying](https://knaaptime.com/urban_analysis/05_access/simple_access.html#using-the-z-dimension) with pydeck lately, which, basically, requires a column of colors on the df. Looking a bit closer, that's how most (all?) of the viz libraries work, so this is a shot at refactoring geopandas's logic for generating colors.
> 
> I can add tests, but wanted to open first to see if folks find this valuable and if it should live here

This is definitely useful functionality.

I'm not sure that mc is the place for this, for a couple of reasons. First, it makes matplotlib a hard dependency. Second, it seems to consume mc, rather than extend mc , for a particular use case. 

I can see a few options:

Keep in it mc, but move it into a utility module that makes matplotlib a soft dependency.

Keep it as suggested with the hard dependency

Don't include it in mc but have it be part of a downstream package.

Other?

**Comment by knaaptime** [comment]
yeah agree. I'm happy to stick it downstream, but when @martinfleis and I discussed briefly on the last pysal call, we thought it might be useful more broadly (e.g. for geopandas to consume, since its coloring logic is currently a bit arcane).

would be easy to make matplotlib a module-level import if we want to go that route

**Comment by martinfleis** [comment]
I've discussed this with @sjsrey in Basel and came to a conclusion that it won't help geopandas at this point. We are planning on refactoring (or at least having a deep dive) of the plotting code for 1.1 or 1.2 though, so it may become more useful then. The issue is that we will always have mapclassify as optional dependency only and when no binning is applied we need to deal with missing values anyway.

**Comment by knaaptime** [comment]
cool. nice work on 1.0 btw!

i still think this has value here, e.g. for legendgram and such (would also give a logical place to live alongside #173), but lmk what you think. Happy to put elsewhere if you prefer

**Comment by martinfleis** [comment]
I am perfectly fine with keeping it here as I can see it can be useful. Just wanted to close the discussion about usage of this in geopandas.

**Comment by knaaptime** [comment]
I vote to put this into a util module with mc as a soft dep. It's not central to classification, but often useful *alongside* classification. 

So it's really useful to keep around with the same group of tools, but shouldn't induce a new dependency

File: mapclassify/util.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        raise ValueError("alpha must be in the range [0,1]")
+    if not pd.api.types.is_list_like(nan_color) and not len(nan_color) == 4:
+        raise ValueError("`nan_color` must be list-like of 4 values: (R,G,B,A)")
+
+    # only operate on non-NaN values
+    v = pd.Series(values, dtype=object)
+    legit_indices = v[~v.isna()].index.values
+
+    # transform (non-NaN) values into class bins
+    bins = classify(v.dropna().values, scheme=classifier, k=k).yb
```
Comment (martinfleis) [inline mapclassify/util.py:51]: Shall we not require `k` and pass through all `kwargs` to support all classifiers and their options?

Comment (knaaptime) [reply mapclassify/util.py:51]: oh, dur. good point

File: mapclassify/util.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    values,
+    classifier="quantiles",
+    cmap="viridis",
+    alpha=1,
+    nan_color=[255, 255, 255, 255],
+    **kwargs
+):
+    """Convert array of values into RGBA colors using a colormap and classifier.
+
+    Parameters
```
Comment (martinfleis) [inline mapclassify/util.py:14]: kwargs should probably be mentioned here.

**Review by martinfleis — APPROVED** [review]
If you make pre-commit happy lgtm.

File: mapclassify/util.py (hunk 3 above)
Comment (knaaptime) [reply mapclassify/util.py:14]: cool, i'm always unsure what to do in the docstrings when there's a catchall

Comment (martinfleis) [reply mapclassify/util.py:14]: I always mention where are they passed to.

**Pull request merged**